source/compilation/builtins/CoverageFuncs.cpp

```diff
@@ -0,0 +1,107 @@
+//------------------------------------------------------------------------------
+// CoverageFuncs.cpp
+// Coverage control functions
+//
+// File is under the MIT license; see LICENSE for details
+//------------------------------------------------------------------------------
+#include "slang/binding/MiscExpressions.h"
+#include "slang/binding/SystemSubroutine.h"
+#include "slang/compilation/Compilation.h"
+#include "slang/diagnostics/SysFuncsDiags.h"
+#include "slang/symbols/InstanceSymbols.h"
+#include "slang/syntax/AllSyntax.h"
+
+namespace slang::Builtins {
+
+class CoverageNameOrHierFunc : public SystemSubroutine {
+public:
+
+    CoverageNameOrHierFunc(const std::string& name, const Type& returnType, unsigned int nameOrHierIndex,
+                           size_t requiredArgs = 0, const std::vector<const Type*>& argTypes = {}) :
+        SystemSubroutine(name, SubroutineKind::Function),
+        argTypes(argTypes), returnType(&returnType),
+        nameOrHierIndex(nameOrHierIndex), requiredArgs(requiredArgs) {
+        ASSERT(requiredArgs <= argTypes.size());
+        ASSERT(nameOrHierIndex <= argTypes.size());
+    };
+
+    virtual const Expression& bindArgument(size_t argIndex, const BindContext& context,
+                                           const ExpressionSyntax& syntax, const Args& args) const final {
+        const BindContext* ctx = &context;
+
+        if (argIndex >= argTypes.size())
+            return SystemSubroutine::bindArgument(argIndex, *ctx, syntax, args);
+
+        if (argIndex == nameOrHierIndex && NameSyntax::isKind(syntax.kind)) {
+            return HierarchicalReferenceExpression::fromSyntax(context.getCompilation(),
+               syntax.as<NameSyntax>(), context);
+        }
+
+        return Expression::bindArgument(*argTypes[argIndex], ArgumentDirection::In, syntax, *ctx);
+    }
+
+    const Type& checkArguments(const BindContext& context, const Args& args,
+                               SourceRange range, const Expression*) const {
+        auto& comp = context.getCompilation();
+        if (!checkArgCount(context, false, args, range, requiredArgs, argTypes.size()))
+            return comp.getErrorType();
+
+        auto& arg = args[nameOrHierIndex];
+        if (arg->kind == ExpressionKind::HierarchicalReference) {
+            auto& sym = *arg->as<HierarchicalReferenceExpression>().symbol;
+            if (sym.isValue()) {
+                auto& type = sym.as<ValueSymbol>().getType();
+                if (!type.isString()) {
+                    context.addDiag(diag::BadSystemSubroutineArg, arg->sourceRange)
+                        << type << kindStr();
+                    return comp.getErrorType();
+                }
+            } else if (sym.kind != SymbolKind::Instance || !sym.as<InstanceSymbol>().isModule()) {
+                if (!context.scope->isUninstantiated())
+                    context.addDiag(diag::ExpectedModuleName, arg->sourceRange);
```

Comment:
It's subtle, but I think the ExpectedModuleInstance is a better diagnostic here.